File: algorithms/sort-algorithms/bubble-sort/bubble-sort.js
// Worst case performance of O(n^2) for comparisons and O(n^2) for swaps
// Best case performance of O(n^2) for comparisons and O(1) for swaps

function bubbleSort(arr) {
  let swapped = false;
  for (let i = 0; i < arr.length; i++) {
    for (let j = 0; j < arr.length; j++) {
      if (arr[j] > arr[j + 1]) {
        [arr[j], arr[j + 1]] = [arr[j + 1], arr[j]];
        swapped = true;
      }
    }
    if (!swapped) {
      break;
    } else {
      swapped = false;
    }
  }
  return arr;
}

// Expected return of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
console.log(bubbleSort([9, 10, 2, 4, 5, 3, 6, 8, 7, 1]));

File: algorithms/sort-algorithms/bubble-sort/bubble_sort.py
def bubble_sort(arr):
    for i in range(0, len(arr)):
        # Boolean flag to check if current iteration has swapped anything
        swapped = False
        # Loops over every time every time
        for j in range(0, len(arr) - 1):
            # Swap elements if j is lesser than j + 1
            if arr[j] > arr[j + 1]:
                # Swap elements
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                # Indicate via boolean flag that change has occured
                swapped = True
        # End all loops if not swaps have happend on current iteration
        if not swapped:
            break
    return arr
